File: web/src/shared/api/skillHub.ts
import { apiDelete, apiGet, apiPost } from "./client";

export type HubSkillVersion = {
  version: string;
  changelog: string;
  sha256: string;
  size_bytes: number;
  scan_status: string;
  scan_warnings: string[];
  published_at: string;
};

export type HubSkillSummary = {
  namespace: string;
  slug: string;
  name: string;
  summary: string;
  tags: string[];
  latest_version: string;
  downloads: number;
  stars: number;
  rating_average: number;
  rating_count: number;
  updated_at: string;
};

export type HubSkillDetail = HubSkillSummary & {
  description: string;
  visibility: string;
  author_username: string;
  versions: HubSkillVersion[];
};

export type InstalledHubSkill = {
  name: string;
  namespace: string;
  slug: string;
  version: string;
  sha256: string;
  installed_at: string;
};

type HubSkillListResponse = {
  data?: {
    items: HubSkillSummary[];
    total: number;
    page: number;
    page_size: number;
  } | null;
};

type HubSkillDetailResponse = {
  data?: HubSkillDetail | null;
};

type InstalledHubSkillsResponse = {
  data?: { items: InstalledHubSkill[] } | null;
};

type InstallHubSkillResponse = {
  data?: {
    installed: InstalledHubSkill;
    updated: boolean;
  } | null;
};

const HUB_PATH = "/api/skill-hub";

export function listHubSkills(params: { q?: string; sort?: string } = {}) {
  const query = new URLSearchParams();
  if (params.q) query.set("q", params.q);
  if (params.sort) query.set("sort", params.sort);
  return apiGet<HubSkillListResponse>(`${HUB_PATH}/skills?${query}`);
}

export function getHubSkill(namespace: string, slug: string) {
  return apiGet<HubSkillDetailResponse>(
    `${HUB_PATH}/skills/${encodeURIComponent(namespace)}/${encodeURIComponent(slug)}`,
  );
}

export function listInstalledHubSkills() {
  return apiGet<InstalledHubSkillsResponse>(`${HUB_PATH}/installed`);
}

export function installHubSkill(payload: { namespace: string; slug: string; version?: string }) {
  return apiPost<InstallHubSkillResponse>(`${HUB_PATH}/install`, payload);
}

export function uninstallHubSkill(name: string) {
  return apiDelete(`${HUB_PATH}/installed/${encodeURIComponent(name)}`);
}

File: schema/skill_hub.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class HubSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")


class HubSkillVersionSchema(HubSchema):
    version: str
    changelog: str
    sha256: str
    size_bytes: int
    scan_status: str
    scan_warnings: list[str]
    published_at: datetime


class HubSkillSummarySchema(HubSchema):
    namespace: str
    slug: str
    name: str
    summary: str
    tags: list[str]
    latest_version: str
    downloads: int
    stars: int
    rating_average: float
    rating_count: int
    updated_at: datetime


class HubSkillDetailSchema(HubSkillSummarySchema):
    description: str
    visibility: str
    author_username: str
    versions: list[HubSkillVersionSchema]
    starred: bool = False
    my_rating: int | None = None


class HubSkillListSchema(HubSchema):
    items: list[HubSkillSummarySchema]
    total: int
    page: int
    page_size: int


class InstalledHubSkillSchema(HubSchema):
    name: str
    namespace: str
    slug: str
    version: str
    sha256: str
    installed_at: datetime


class InstallHubSkillRequest(HubSchema):
    namespace: str = Field(min_length=1, max_length=48)
    slug: str = Field(min_length=1, max_length=64)
    version: str = Field(default="", max_length=64)


class InstallHubSkillResponse(HubSchema):
    installed: InstalledHubSkillSchema
    updated: bool
